### PR TITLE
Altair review

### PR DESCRIPTION
**Pull request by djrtwo**
* add some missing beacon-chain function comments
* Reorder attestation logic to be in order [source, target, head] throughout
* define `is_{attestation_field}` as having to have the parent `is_{attestation_field}` correct, thus enforcing chain coherence any time those booleans are used
* add sync committee test for empty participants

### CODE REVIEW

File: specs/altair/beacon-chain.md [hunk 1]
```diff
@@ -455,22 +470,22 @@ def process_attestation(state: BeaconState, attestation: Attestation) -> None:
         justified_checkpoint = state.previous_justified_checkpoint
 
     # Matching roots
-    is_matching_head = data.beacon_block_root == get_block_root_at_slot(state, data.slot)
     is_matching_source = data.source == justified_checkpoint
-    is_matching_target = data.target.root == get_block_root(state, data.target.epoch)
+    is_matching_target = is_matching_source and data.target.root == get_block_root(state, data.target.epoch)
+    is_matching_head = is_matching_target and data.beacon_block_root == get_block_root_at_slot(state, data.slot)
```
Comment (protolambda) [inline specs/altair/beacon-chain.md:475]: The condition changed from just a head block-root check, to a is-target + head block root check. And the if statements below changed to then not do the target check again. I think we were testing good head votes with bad target votes, but if not, I'll try add this after the specs call.

File: specs/altair/beacon-chain.md [hunk 2]
```diff
@@ -455,22 +470,22 @@ def process_attestation(state: BeaconState, attestation: Attestation) -> None:
...
 
     # Participation flag indices
     participation_flag_indices = []
-    if is_matching_head and is_matching_target and state.slot == data.slot + MIN_ATTESTATION_INCLUSION_DELAY:
-        participation_flag_indices.append(TIMELY_HEAD_FLAG_INDEX)
     if is_matching_source and state.slot <= data.slot + integer_squareroot(SLOTS_PER_EPOCH):
         participation_flag_indices.append(TIMELY_SOURCE_FLAG_INDEX)
     if is_matching_target and state.slot <= data.slot + SLOTS_PER_EPOCH:
         participation_flag_indices.append(TIMELY_TARGET_FLAG_INDEX)
+    if is_matching_head and state.slot == data.slot + MIN_ATTESTATION_INCLUSION_DELAY:
```
Comment (protolambda) [inline specs/altair/beacon-chain.md:487]: Yay for source>target>head ordering of logic.

File: specs/altair/beacon-chain.md [hunk 3]
```diff
@@ -311,10 +320,16 @@ def get_base_reward_per_increment(state: BeaconState) -> Gwei:
...
 
-*Note*: The function `get_base_reward` is modified with the removal of `BASE_REWARDS_PER_EPOCH`.
+*Note*: The function `get_base_reward` is modified with the removal of `BASE_REWARDS_PER_EPOCH` and the use of increment based accounting.
 
 ```python
 def get_base_reward(state: BeaconState, index: ValidatorIndex) -> Gwei:
+    """
+    Return the base reward for the validator defined by ``index`` with respect to the current ``state``.
+
+    Note: A validator can optimally earn one base reward per epoch over a long time horizon.
```
Comment (protolambda) [inline specs/altair/beacon-chain.md:330]: We should make a note of the difference with phase0. It's still the same amount of ETH, but the base reward naming changed from 4 (`BASE_REWARDS_PER_EPOCH`) to just 1 big one per epoch + weight logic.

Comment (protolambda) [reply specs/altair/beacon-chain.md:330]: Also, the precision changed:

Phase0:
```
effective_balance * BASE_REWARD_FACTOR // integer_squareroot(total_balance) // BASE_REWARDS_PER_EPOCH
```

Altair:
```
EFFECTIVE_BALANCE_INCREMENT * BASE_REWARD_FACTOR // integer_squareroot(get_total_active_balance(state))
```

Napkin math: assuming an effective balance of ~32, that's 5 bits of extra precision that don't get rounded down. Then it divides by `BASE_REWARDS_PER_EPOCH = 4`, 2 bits less, but it repeats 4 times, so the effect may not be as big there.

The higher the precision, the less the base reward gets "rounded down". So those ~5 bits = 32 gwei (pessimistic maximum) lost per validator per attestation when forking to Altair. In the grand scheme of things it's a tiny rounding thing, but it's about validator rewards, so I think we want to get it right.

File: specs/altair/beacon-chain.md (hunk 1 above)
Comment (djrtwo) [reply specs/altair/beacon-chain.md:476]: This is not the case. See here:

https://github.com/ethereum/eth2.0-specs/pull/2374/files#diff-76363249532da52344c30ccc6864859e2a74ad83130d0d640033cbbe4ff0a5beL468

We used to write `is_matching_head and is_matching_target` whereas this changes `is_matching_head` to include `is_matching_target` to remove the footgun of using head alone somewhere by accident

Comment (djrtwo) [reply specs/altair/beacon-chain.md:476]: EDIT
I misread your statement. You are correct in your assessment and we almost certainly have such tests

Comment (protolambda) [reply specs/altair/beacon-chain.md:476]: > And the if statements below changed to then not do the target check again.

That's what I said, "to not check the target *again*", after it being checked as part of `is_matching_head`. Anyway, it looks consistent, but we need to check testing.

Comment (protolambda) [reply specs/altair/beacon-chain.md:476]: Confirmed we have tests for it. Bad target while a valid head vote is processed, with multiple cases for inclusion delay too: https://github.com/ethereum/eth2.0-specs/blob/0c94be420405380f104347e643500f4613ab8f9c/tests/core/pyspec/eth2spec/test/phase0/block_processing/test_process_attestation.py#L498

File: specs/altair/beacon-chain.md (hunk 3 above)
Comment (protolambda) [reply specs/altair/beacon-chain.md:330]: I'll make an issue to review the precision of Altair rewards, but I don't think it's significant enough to block this PR, especially since rewards change in other bigger ways.

**Review by protolambda — APPROVED** [review]
LGTM

File: specs/altair/beacon-chain.md [hunk 4]
```diff
@@ -311,10 +320,17 @@ def get_base_reward_per_increment(state: BeaconState) -> Gwei:
 
 #### `get_base_reward`
 
-*Note*: The function `get_base_reward` is modified with the removal of `BASE_REWARDS_PER_EPOCH`.
+*Note*: The function `get_base_reward` is modified with the removal of `BASE_REWARDS_PER_EPOCH` and the use of increment based accounting.
```
Comment (ralexstokes) [inline specs/altair/beacon-chain.md:323]: tbh i'm not really sure what "increment based" means but i don't think it blocks this PR

**Review by ralexstokes — APPROVED** [review]
lgtm!

i left a comment about "increment based" accounting bc i think there is a more appropriate term but doesn't really block this PR
